fvtest/tril/tril/parser.cpp

```diff
@@ -0,0 +1,560 @@
+                        if (failed(state)) {
+                            munchedInput += *inputPosn;
+                        }
+                        throw LexerFailure("ERROR: Simplified maximal munch failed on input: " + munchedInput);
+                    }
+                }
+            }
+            return result;
+        }
+};
+
+
+/* Scan the given input and filtered useful tokens
+ * 
+ * @param input The whole tril input
+ * @return The vector of useful tokens
+ */
+ 
+std::vector<Token> scan(const std::string &input) {
+    static ASTstates rules;
+    std::vector <Token> tokens;
+    try {
+        tokens = rules.simplifiedMaximalMunch(input);
+    } catch (LexerFailure &f) {
+        std::cerr << f.what() << std::endl;
+    }
+    std::vector <Token> newTokens;
+    for (int i = 0; i < tokens.size(); i++) {
+        if (tokens[i].getType() != Token::Type::COMMENT && tokens[i].getType() != Token::Type::WHITESPACE) {
+			newTokens.push_back(tokens[i]);
+        }
+    }
+    return newTokens;
+}
+
+
+std::ostream &operator<<(std::ostream &out, const Token &tok) {
+    switch(tok.getType()) {
+        case Token::ID: out << "ID"; break;
+        case Token::STRING_: out << "STRING_"; break;
+        case Token::INT: out << "INT"; break;
+        case Token::DOUBLE_: out << "DOUBLE_"; break;
+        case Token::AT:  out << "AT"; break;
+        case Token::ZERO:  out << "ZERO"; break;
+        case Token::QUOT:  out << "QUOT"; break;
+        case Token::LPAREN:  out << "LPAREN"; break;
+        case Token::RPAREN:  out << "RPAREN"; break;
+        case Token::LPAREN_SQ:  out << "LPAREN_SQ"; break;
+        case Token::RPAREN_SQ:  out << "RPAREN_SQ"; break;
+        case Token::EQUALS: out << "EQUALS"; break;
+        case Token::COMMA: out << "COMMA"; break;
+        case Token::HEXINT:  out << "HEXINT"; break;
+        default: throw LexerFailure("ERROR: Failed to determine input: " + tok.getValue());
+    }
+    out << " " << tok.getValue();
+    return out;
+}
+
+class TokenIter {
+    private:
+        std::vector <Token>::iterator it;
+        std::vector <Token>::iterator end;
+    public:
+        TokenIter(std::vector <Token>::iterator it, std::vector <Token>::iterator end)
+            : it(it), end(end) {}
+        TokenIter& operator++() {
+            ++it;
+            return *this;
+        }
+        bool isEnd() { return it == end; }
+        Token peek() {
+            auto temp = it;
+            ++temp;
+            return *temp;
+        }
+        Token& operator * () { return *it; }
+        Token* operator -> () { return &*it; }
+};
+
+/* Grammer rules:
+ * nodeList: node | nodeList node
+ * node: '(' IDENTIFIER ')' | '(' IDENTIFIER nodeList ')' | '(' IDENTIFIER argList ')'	| '(' IDENTIFIER argList nodeList ')'
+ * argList: arg | argList arg
+ * arg: value | IDENTIFIER '=' value | IDENTIFIER '=' '[' valueList ']'
+ * valueList: value | valueList ',' value
+ * value: INTEGER | DOUBLE | STRING | IDENTIFER
```

Comment:
Can I use below comment as instead?
```
nodeList: epsilon | node | nodeList node
node: '(' IDENTIFIER argList nodeList ')'
argList: epsilon | arg | argList arg
```

Reply:
I would suggest using another character instead of explicitly saying "epsilon." For example, you can use "." and just put a note somewhere saying that you're using it instead of epsilon: 

```
nodeList: nodeList node | .
node: '(' IDENTIFIER argList nodeList ')' 
argList: argList arg | .

(note: "." is being used to designate epsilon/null productions)
```